### PR TITLE
Support datetime2 cast and rounding in datetime2 TRY_CONVERT

### PR DESCRIPTION
**Pull request by Jakeowen1**
### Description

Currently try_convert to datetime2 cannot handle certain cases. This commit changes the implementation to use cast, and also extends the datetime2 implementation to cover try_cast. It reverts the changes of the previous commit to rely on cast rather than the datetime_to_string function
 
Task: Babel-3614

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by Sairakan — APPROVED** [review]

File: contrib/babelfishpg_tsql/sql/sys_function_helpers.sql [hunk 1]
```diff
@@ -1639,10 +1639,6 @@ BEGIN
             v_day := substr(v_datestring, 7, 2);
             v_month := substr(v_datestring, 5, 2);
             v_year := substr(v_datestring, 1, 4);
-        ELSE
```
Comment (timchang514) [inline contrib/babelfishpg_tsql/sql/sys_function_helpers.sql:1642]: Do we need to replace this ELSE statement with something? Like a default value?

File: test/JDBC/expected/BABEL-3614-vu-verify.out [hunk 2]
```diff
@@ -202,8 +204,9 @@ SELECT BABEL_3614_vu_prepare_f7()
 GO
 ~~START~~
 datetime2
-<NULL>
-~~END~~
```
Comment (timchang514) [inline test/JDBC/expected/BABEL-3614-vu-verify.out:206]: Just want to make sure, is this the same behavior as SQL Server - do they return NULL or throw an error in this case?

Comment (timchang514) [reply test/JDBC/expected/BABEL-3614-vu-verify.out:206]: For this case and all the other ones like it in this file*

File: contrib/babelfishpg_tsql/sql/sys_function_helpers.sql (hunk 1 above)
Comment (Jakeowen1) [reply contrib/babelfishpg_tsql/sql/sys_function_helpers.sql:1642]: This is how the function had it before I made the changes in my last pr so im just restoring it to the original version. If it makes sense to keep this change I can put it back

File: test/JDBC/expected/BABEL-3614-vu-verify.out (hunk 2 above)
Comment (Jakeowen1) [reply test/JDBC/expected/BABEL-3614-vu-verify.out:206]: Yes this is how sql server handles it. If conversion to that type is prohibited it throws an error. If conversion to that type is allowed but the conversion cannot be completed it returns null. Will run through all of the test cases to confirm

Comment (Jakeowen1) [reply test/JDBC/expected/BABEL-3614-vu-verify.out:206]: Yeah so everything behaves the same aside from some of the rounding which just has to do with our difference in representation. Errors are thrown in the same cases though

**Review by AvantikaDG — APPROVED** [review]

**Comment by AvantikaDG** [comment]
Let's change the PR title to something like "Support datetime2 cast and rounding in datetime2 TRY_CONVERT". You can mention in the description that you reverted the previous commit Also does this cover BABEL-3646?

**Comment by Jakeowen1** [comment]
> Let's change the PR title to something like "Support datetime2 cast and rounding in datetime2 TRY_CONVERT". You can mention in the description that you reverted the previous commit Also does this cover BABEL-3646?

This commit does not cover babel-3646 because that involves more datatypes than just datetime2. This commit just fixes the datetime2 issues but also enables support for try_cast datetime2

**Pull request merged**